File: 332_Reconstruct_Itinerary.cpp
class Solution {
private:
    unordered_map<string, multiset<string>> targets;
    vector<string> path;
    void dfs(string departure) {
        while (targets[departure].size()) {
            string arrival = *targets[departure].begin();
            targets[departure].erase(targets[departure].begin());
            dfs(arrival);
        }
        path.push_back(departure);
    }
public:
    vector<string> findItinerary(vector<pair<string, string>> tickets) {
        for (auto t : tickets) targets[t.first].insert(t.second);
        dfs("JFK");
        return vector<string> (path.rbegin(), path.rend());
    }
};

File: 332_Reconstruct_Itinerary.py
from collections import defaultdict
import heapq
class Solution:
    def dfs(self, dpt):
        while len(self.targets[dpt]):
            arv = heapq.heappop(self.targets[dpt])
            self.dfs(arv)
        self.res.append(dpt)
        
        
    def findItinerary(self, tickets: List[List[str]]) -> List[str]:
        self.targets = defaultdict(list)
        self.res = []
        for dpt, arv in tickets:
            heapq.heappush(self.targets[dpt], arv)
        self.dfs('JFK')
        return self.res[::-1]
